### PR TITLE
Test task

### PR DESCRIPTION
**Pull request by natasha-vardo**

### CODE REVIEW

File: app/code/local/PA/Access/Block/Product/Compare/List.php [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+class PA_Access_Block_Product_Compare_List extends Mage_Catalog_Block_Product_Compare_List
+{
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Block/Product/Compare/List.php:5]: line

File: app/code/local/PA/Access/Block/Product/List.php [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+class PA_Access_Block_Product_List extends Mage_Catalog_Block_Product_List
+{
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Block/Product/List.php:5]: line

File: app/code/local/PA/Access/Block/Product/View.php [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+class PA_Access_Block_Product_View extends Mage_Catalog_Block_Product_View
+{
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Block/Product/View.php:5]: line

File: app/code/local/PA/Access/Helper/Data.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+{
+    /**
+     * @param array $productCategory
+     *
+     * @return bool
+     */
+    public function isRestrictionCategory(array $productCategory): bool
+    {
+        $settingsRestrictionCategory = Mage::getStoreConfig('access/settings/restriction_category');
+        $restrictionCategory = explode(",", $settingsRestrictionCategory);
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Helper/Data.php]: ','

File: app/code/local/PA/Access/Helper/Data.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        $dateTimeNow = $dateTime->format('Y-m-d H:i:s');
+
+        $currentDateTime = $dateTimeNow;
+        $startDateTime = sprintf('%s %s', $dateNow, $startRestriction);
+
+        if ($startRestriction <= $endRestriction) {
+            $endDateTime = sprintf('%s %s', $dateNow, $endRestriction);
+            $previousDate = $dateTime->modify('-1 day')->format('Y-m-d');
+            $previousDayEnd = sprintf('%s %s', $previousDate, $endRestriction);
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Helper/Data.php]: line

File: app/code/local/PA/Access/Helper/Data.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+     *
+     * @throws Exception
+     */
+    public function getRestriction(string $startRestriction, string $endRestriction): array
+    {
+        $dateTime = new DateTime('now');
+        $dateNow = $dateTime->format('Y-m-d');
+        $dateTimeNow = $dateTime->format('Y-m-d H:i:s');
+
+        $currentDateTime = $dateTimeNow;
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Helper/Data.php]: Why?

File: app/code/local/PA/Access/Helper/Data.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            $previousDate = $dateTime->modify('-1 day')->format('Y-m-d');
+            $previousDayEnd = sprintf('%s %s', $previousDate, $endRestriction);
+
+        } else {
+            $endDate = $dateTime->modify('+1 day')->format('Y-m-d');
+            $endDateTime = sprintf('%s %s', $endDate, $endRestriction);
+            $previousDayEnd = sprintf('%s %s', $dateNow, $endRestriction);
+        }
+
+        $dateArray = [
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Helper/Data.php]: return []

File: app/code/local/PA/Access/Model/Adminhtml/System/Config/Source/Shopcategory.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+class PA_Access_Model_Adminhtml_System_Config_Source_Shopcategory
+{
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Adminhtml/System/Config/Source/Shopcategory.php]: line

File: app/code/local/PA/Access/Model/Adminhtml/System/Config/Source/Shopcategory.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * @return array
+     */
+    public function toOptionArray(): array
+    {
+        $categories = Mage::getModel('catalog/category')->getCollection()->load();
+        $categoriesIdArray = [];
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Adminhtml/System/Config/Source/Shopcategory.php]: $categoriesId = []

File: app/code/local/PA/Access/Model/Adminhtml/System/Config/Source/Shopcategory.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+            if ($category['level'] !== '0' && $category['level'] !== '1') {
+                $categoriesIdArray[] = [
+                    'value' => $category['entity_id']
+                ];
+            }
+        }
+
+        $options = [];
+
+        foreach ($categoriesIdArray as $id) {
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Adminhtml/System/Config/Source/Shopcategory.php]: id is a bad name. id isn't array.

File: app/code/local/PA/Access/Model/Observer.php [hunk 11]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+class PA_Access_Model_Observer
+{
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Observer.php]: line

File: app/code/local/PA/Access/Model/Observer.php [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * @param int $productId
+     *
+     * @return array|null
+     */
+    public function getCategoryIds(int $productId): array
+    {
+        if (isset($productId)) {
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Observer.php]: when productId is not set?

File: app/code/local/PA/Access/Model/Observer.php [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    public function checkout_cart_save_before(Varien_Event_Observer $observer)
+    {
+        $startRestrictionFieldInfo = Mage::getStoreConfig('access/settings/start_restriction');
+        $endRestrictionFieldInfo = Mage::getStoreConfig('access/settings/end_restriction');
+
+        $settingsRestrictionCategory = Mage::getStoreConfig('access/settings/restriction_category');
+        $restrictionCategory = explode(",", $settingsRestrictionCategory);
+
+        $startRestriction = strtr($startRestrictionFieldInfo, ",", ":");
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Observer.php]: ','

File: app/code/local/PA/Access/Model/Observer.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+        $endRestriction = strtr($endRestrictionFieldInfo, ",", ":");
+
+        $helper = Mage::helper('access');
+        $restriction = $helper->getRestriction($startRestriction, $endRestriction);
+
+        $items = $observer->getEvent()->getData('cart')->getItems();
+
+            foreach ($items as $item) {
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Observer.php]: line

File: app/code/local/PA/Access/Model/Observer.php [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+                $categoryIds = $this->getCategoryIds($item['product_id']);
+
+                foreach ($restrictionCategory as $category) {
+
+                    if (in_array($category, $categoryIds)) {
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Observer.php]: line

File: app/code/local/PA/Access/Model/Observer.php [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+                        if ($restriction['currentDateTime'] >= $restriction['startDateTime']
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Observer.php]: use variable

File: app/code/local/PA/Access/Model/Observer.php [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        $restriction = $helper->getRestriction($startRestriction, $endRestriction);
+
+        $items = $observer->getEvent()->getData('cart')->getItems();
+
+            foreach ($items as $item) {
+
+                $categoryIds = $this->getCategoryIds($item['product_id']);
+
+                foreach ($restrictionCategory as $category) {
+
```
Comment (DutiFruti) [inline app/code/local/PA/Access/Model/Observer.php]: line

File: app/design/frontend/rwd/default/template/pa/addtocart.phtml [hunk 18]
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
Comment (DutiFruti) [inline app/design/frontend/rwd/default/template/pa/addtocart.phtml]: line

File: app/design/frontend/rwd/default/template/pa/addtocart.phtml [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    $_product = $this->getProduct();
```
Comment (DutiFruti) [inline app/design/frontend/rwd/default/template/pa/addtocart.phtml:2]: why _?

File: app/design/frontend/rwd/default/template/pa/addtocart.phtml [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+    $productCategory = $_product->getCategoryIds();
+
+    $helper = Mage::helper('access');
+    $isRestriction = $helper->isRestrictionCategory($productCategory);
+
+    $startRestrictionFieldInfo = Mage::getStoreConfig('access/settings/start_restriction');
+    $endRestrictionFieldInfo = Mage::getStoreConfig('access/settings/end_restriction');
+
+    $startRestriction = strtr($startRestrictionFieldInfo, ",", ":");
```
Comment (DutiFruti) [inline app/design/frontend/rwd/default/template/pa/addtocart.phtml]: ''

File: app/design/frontend/rwd/default/template/pa/addtocart.phtml [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    $endRestriction = strtr($endRestrictionFieldInfo, ",", ":");
+
+    $restriction = $helper->getRestriction($startRestriction, $endRestriction);
+
+    if ($restriction['currentDateTime'] >= $restriction['startDateTime']
+        && $restriction['currentDateTime'] <= $restriction['endDateTime'] && $isRestriction) {
+        echo '<div class="add-to-cart-buttons"><strong>Sorry, you cant buy this product until '.substr($endRestriction, 0, -3).'</strong></div>';
```
Comment (DutiFruti) [inline app/design/frontend/rwd/default/template/pa/addtocart.phtml]: substr($endRestriction, 0, -3)?

File: app/design/frontend/rwd/default/template/pa/cat_access.phtml [hunk 22]
```diff
@@ -0,0 +1,213 @@
+<?php
```
Comment (DutiFruti) [inline app/design/frontend/rwd/default/template/pa/cat_access.phtml:1]: Everything is bad. Logic is better to move :)

File: app/design/frontend/rwd/default/template/pa/cat_access.phtml [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+/**
+ * Product list template
+ *
+ * @see Mage_Catalog_Block_Product_List
+ */
+/* @var $this Mage_Catalog_Block_Product_List */
+?>
+<?php
+$_productCollection=$this->getLoadedProductCollection();
```
Comment (DutiFruti) [inline app/design/frontend/rwd/default/template/pa/cat_access.phtml:17]: _

File: app/design/frontend/rwd/default/template/pa/compare.phtml [hunk 24]
```diff
@@ -0,0 +1,188 @@
+<?php
```
Comment (DutiFruti) [inline app/design/frontend/rwd/default/template/pa/compare.phtml:1]: Everything is bad. Logic is better to move :)

File: app/design/frontend/rwd/default/template/pa/addtocart.phtml (hunk 19 above)
Comment (natasha-vardo) [reply app/design/frontend/rwd/default/template/pa/addtocart.phtml:2]: it is default variable

File: app/design/frontend/rwd/default/template/pa/cat_access.phtml (hunk 23 above)
Comment (natasha-vardo) [reply app/design/frontend/rwd/default/template/pa/cat_access.phtml:17]: it is default variable

File: app/design/frontend/rwd/default/template/pa/addtocart.phtml (hunk 21 above)
Comment (natasha-vardo) [reply app/design/frontend/rwd/default/template/pa/addtocart.phtml]: value from settings comes like "08:00:00", after substr function - "08:00"

**Pull request merged**